compose-lint-checks/src/main/java/slack/lint/compose/ModifierReusedDetector.kt

```diff
@@ -57,123 +63,328 @@ constructor(
-                .filterIsInstance<KtCallExpression>()
-                .filter { it.isUsingModifiers(modifierNames) }
-            )
-            current = current.parent
-          }
+    val modifierReferences = mutableSetOf<PsiElement>()
+    modifier.sourcePsi?.let { modifierReferences.add(it) }
+    // Analyze data flow to get all possible references of modifier parameter
+    method.accept(
+      object : DataFlowAnalyzer(setOf(modifier)) {
+        override fun receiver(call: UCallExpression) {
+          val reference = call.receiver?.skipParenthesizedExprDown()?.sourcePsi
+          if (reference == null || reference is KtDotQualifiedExpression) return
+          modifierReferences += reference
+        }
+
+        override fun argument(call: UCallExpression, reference: UElement) {
+          // if modifier is an argument of the call (i.e. Modifier.then(modifier)), then track
+          track(call)
+          // that call as if it is returning this modifier
+          modifierReferences += reference.sourcePsi ?: return
         }
       }
-      // Any set with more than 1 item is interesting to us: means there is a rule violation
-      .filter { it.size > 1 }
-      // At this point we have all the grouping of violations, so we just need to extract all
-      // individual
-      // items from them as we are no longer interested in the groupings, but their individual
-      // elements
-      .flatten()
-      // We don't want to double report
+    )
+
+    val cfaAst = method.toSimplifiedCfaAst(modifierReferences)
+    val (graph, firstNode) = cfaAst.toCodeFlowGraph()
+    val result = HashSet<GraphNode>()
+    findPathsWithMultipleModifierUses(firstNode, graph, listOf(), result)
+
+    result
+      .map { it.call!! }
+      .filter { callExpression ->
+        callExpression.calleeExpression?.text?.first()?.isUpperCase() == true
+      }
+      .flatMap { callExpression ->
+        (callExpression.valueArgumentList ?: return@flatMap emptySequence())
+          .findChildrenByClass<KtNameReferenceExpression>()
+          .filter { it in modifierReferences }
+      }
       .distinct()
-      .forEach { callExpression ->
+      .forEach { modifierArgument ->
         context.report(
           ISSUE,
-          callExpression,
-          context.getLocation(callExpression),
+          modifierArgument,
+          context.getLocation(modifierArgument),
           ISSUE.getExplanation(TextFormat.TEXT),
         )
       }
   }
 
-  private fun KtCallExpression.isUsingModifiers(modifierNames: List<String>): Boolean =
-    valueArguments.any { argument ->
-      when (val expression = argument.getArgumentExpression()) {
-        // if it's MyComposable(modifier) or similar
-        is KtReferenceExpression -> {
-          modifierNames.contains(expression.text)
+  // Convert UAST into simplified AST, that will be used in code flow analysis
+  private fun UMethod.toSimplifiedCfaAst(references: Set<PsiElement>): CfaBlock {
```

Comment:
Can we pull this custom AST code out to its own file or package? And ideally with a lot more docs? This looks like it's adding quite a bit of complexity to the check on the face of it

Reply:
Sure

Reply:
Didn't extract custom AST into other file, but added a lot more docs and fixed a couple of bugs from previous version